README.rst
Mixcloud.com API access from Python
-----------------------------------

|Build Status| |Coverage Status|

A fork of emillon/mixcloud

This provides a Python API for the http://mixcloud.com website.

Example
-------

.. code:: python

    from mixcloud import Mixcloud
    m = Mixcloud()
    u = m.user('michelplatiniste')
    for c in u.cloudcasts():
        print c.name

Authorization
-------------

You must have an application registered with Mixcloud and a corresponding
client ID and secret.

Authorization is a multi-step process:

1. Generate the authorization URL
2. Redirect the user to the URL to authorize your application
3. Collect the `code` query string parameter from the redirect
4. Exchange the code for an access token

.. code:: python

    o = mixcloud.MixcloudOauth(
        client_id=client_id, client_secret=client_secret,
        redirect_uri=redirect_uri)

    url = o.authorize_url()

    # Next redirect the user to `url`. They will be redirected to your
    # redirect uri with a `code` query string parameter. This must be
    # exchanged with Mixcloud for an access token.

    access_token = o.exchange_token(code)

    # This can then be used for calls that required authorization.

    m = mixcloud.Mixcloud(access_token=access_token)
    print(m.me())

Optionally you can store an access token in your `.netrc` file and it
will be used automatically. The machine name must be `mixcloud-api`.

E.g.::

    machine mixcloud-api
    password my_access_token

Uploading
---------

It is possible to use this module to upload cloudcasts. In order to do
that you need to be authenticated. To do that, provide an API token to
the constructor.

.. code:: python

    m = mixcloud.Mixcloud(access_token=access_token)
    cc = Cloudcast(...)
    with open(mp3_path) as mp3:
        r = m.upload(cc, mp3)

YML file support
----------------

It is possible to represent cloudcasts as YAML files. See
``example.yml``.

The relevant keys are:

+----------+-------------------+
| Key      | Type              |
+==========+===================+
| title    | String            |
+----------+-------------------+
| desc     | String            |
+----------+-------------------+
| tags     | List of strings   |
+----------+-------------------+
| tracks   | List of tracks    |
+----------+-------------------+

Each track is a dict with the following keys:

+----------+-----------+
| Key      | Type      |
+==========+===========+
| start    | Integer   |
+----------+-----------+
| artist   | String    |
+----------+-----------+
| track    | String    |
+----------+-----------+

You can leverage YAML syntax for the "start" field: ``2:54`` will be
parsed directly as 174 (then number of seconds).

It is possible to parse such a file with:

.. code:: python

    with open(yml_path) as yml:
        cc = mixcloud.Cloudcast.from_yml(yml, None)

Mocking
-------

A mock server is provided for testing purposes in ``mixcloud.mock``.

.. |Build Status| image:: https://img.shields.io/travis/emillon/mixcloud/master.svg
   :target: http://travis-ci.org/emillon/mixcloud
.. |Coverage Status| image:: https://img.shields.io/coveralls/emillon/mixcloud/master.svg
   :target: https://coveralls.io/r/emillon/mixcloud


HISTORY.rst
Changelog
=========

unreleased
----------

0.6.0
-----
**release date:** 2021-04-15

* Add playlist functionality (#3, #4)


0.5.0
-----
**release date:** 2021-04-26

* add Git Actions CI
* Refactor to Python3.7+, abandon Python 2 (#1, #2)

0.4.0
-----
**release date:** 2015-10-27

* Switch to semantic versioning.
* Include access token when calling /me (#10)
* Add a helper class for OAuth (#11)
* Support access token stored in netrc (#9, #12)

0.0.3
-----
**release date:** 2015-03-15

* Pictures can be uploaded and retrieved (#6)
* Include LICENSE in tarball (#8)

0.0.2
-----
**release date:** 2015-01-28

* Switch from requirements.txt to install_requires (#7)
* Add HISTORY.rst to MANIFEST.in

0.0.1
-----
**release date:** 2015-01-28

* Initial release


tests.py
import csv
import datetime
import io
import unittest

import dateutil.tz
import httpretty

import mixcloud
from mixcloud.mock import MockServer, parse_headers, parse_multipart

from unittest import mock
from urllib.parse import parse_qs, urlsplit


def parse_tracklist(s):
    s = s.split('\n')[1:-1]
    reader = csv.reader(s, delimiter='|', quoting=csv.QUOTE_NONE)
    t = [mixcloud.Section(int(l[0]),
                          mixcloud.Track(l[1].strip(),
                                         mixcloud.Artist(None,
                                                         l[2].strip()
                                                         )
                                         )
                          )
         for l in reader]
    return t


afx = mixcloud.Artist('aphex-twin', 'Aphex Twin')
spartacus = mixcloud.User('spartacus', 'Spartacus')
partytime = mixcloud.Cloudcast(
    'party-time', 'Party Time',
    parse_tracklist("""
       0 | Samurai (12" Mix)              | Jazztronik
     416 | Refresher                      | Time of your life
     716 | My time (feat. Crystal Waters) | Dutch
    1061 | Definition of House            | Minimal Funk
    1500 | I dont know                    | Mint Royale
    1763 | Thrill Her                     | Michael Jackson
    2123 | Happy (feat.Charlise)          | Elio Isola
    2442 | Dancin                         | Erick Morillo et al
    2738 | All in my head                 | Kosheen
    """),
    ['Funky house', 'Funk', 'Soul'],
    'Bla bla',
    spartacus,
    datetime.datetime(2009, 8, 2, 16, 55, 1, tzinfo=dateutil.tz.tzutc()),
)
lambiance = mixcloud.Cloudcast(
    'lambiance', "L'ambiance",
    parse_tracklist("""
     10 | As Serious As Your Life                     | Four Tet
     20 | Dynamic Symmetry                            | BT
     30 | Vessel                                      | Jon Hopkins
     40 | Vordhosbn                                   | Aphex Twin
     50 | Colour Eye                                  | Jon Hopkins
     60 | Flite                                       | The Cinematic Orchestra
     70 | Altibzz                                     | Autechre
     80 | Untitled [SAW2 CD1 Track1] (Four Tet remix) | Aphex Twin
     90 | Angelica                                    | Lamb
    100 | Quixotic                                    | Spartacus
    110 | Monday - Paracetamol                        | Ulrich Schnauss
    120 | Aquarius                                    | Boards of Canada
    130 | Channel 1 Suite                             | The Cinematic Orchestra
    """),
    ['Idm', 'Originals', 'Ambient'],
    'Bla bla bla',
    spartacus,
    datetime.datetime(2010, 3, 11, 21, 53, 8, tzinfo=dateutil.tz.tzutc()),
)


class TestMixcloud(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        httpretty.enable()

    def setUp(self):
        self.client_id = 'qwerty'
        self.client_secret = 'ytrewq'
        self.redirect_uri = 'http://localhost/mixcloud-callback'
        self.o = mixcloud.MixcloudOauth(
            client_id=self.client_id, client_secret=self.client_secret,
            redirect_uri=self.redirect_uri)
        self.m = mixcloud.Mixcloud()
        httpretty.reset()
        self.mc = MockServer()

    def testArtist(self):
        self.mc.register_artist(afx)
        resp = self.m.artist('aphex-twin')
        self.assertEqual(resp.name, 'Aphex Twin')

    def testUser(self):
        self.mc.register_user(spartacus)
        resp = self.m.user('spartacus')
        self.assertEqual(resp.name, 'Spartacus')

    def testCloudcast(self):
        self.mc.register_cloudcast(spartacus, partytime)
        resp = self.m.user('spartacus')
        cc = resp.cloudcast('party-time')
        self.assertEqual(cc.name, 'Party Time')
        secs = cc.sections()
        self.assertEqual(len(secs), 9)
        sec = secs[1]
        self.assertEqual(sec.start_time, 416)
        self.assertEqual(sec.track.name, 'Refresher')
        self.assertEqual(sec.track.artist.name, 'Time of your life')
        self.assertEqual(cc.tags, ['Funky house', 'Funk', 'Soul'])
        self.assertEqual(cc.description(), 'Bla bla')
        self.assertTrue(cc.picture().startswith('http'))
        self.assertEqual(cc.user.key, 'spartacus')
        self.assertEqual(cc.user.name, 'Spartacus')

    def testCloudcasts(self):
        self.mc.register_cloudcast(spartacus, partytime)
        resp = self.m.user('spartacus')
        ccs = resp.cloudcasts()
        self.assertEqual(len(ccs), 1)
        cc = ccs[0]
        self.assertEqual(cc.name, 'Party Time')

    def testLogin(self):
        self.mc.i_am(spartacus)
        user = self.m.me()
        self.assertEqual(user.key, spartacus.key)
        self.assertEqual(user.name, spartacus.name)

    def testUpload(self):
        self.mc.i_am(spartacus)

        def upload_callback(request, uri, headers):
            data = parse_multipart(request.body)
            self.assertIn('mp3', data)
            self.assertIn('picture', data)
            name = data['name']
            key = mixcloud.slugify(name)
            sections, tags = parse_headers(data)
            description = data['description']
            me = self.m.me()
            created_time = datetime.datetime.now()
            cc = mixcloud.Cloudcast(key, name, sections, tags,
                                    description, me, created_time)
            self.mc.register_cloudcast(me, cc)
            return (200, headers, '{}')

        self.mc.handle_upload(upload_callback)
        mp3file = io.StringIO(u'\x00' * 30)
        picturefile = io.StringIO(u'\x00' * 30)
        r = self.m.upload(partytime, mp3file, picturefile=picturefile)
        self.assertEqual(r.status_code, 200)
        me = self.m.me()
        cc = me.cloudcast('party-time')
        self.assertEqual(cc.name, 'Party Time')
        secs = cc.sections()
        self.assertEqual(len(secs), 9)
        sec = secs[3]
        self.assertEqual(sec.start_time, 1061)
        self.assertEqual(sec.track.name, 'Definition of House')
        self.assertEqual(sec.track.artist.name, 'Minimal Funk')
        self.assertEqual(cc.tags, ['Funky house', 'Funk', 'Soul'])
        self.assertEqual(cc.description(), 'Bla bla')

    def testCloudcastsSection(self):
        self.mc.register_cloudcast(spartacus, partytime)
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        cc = ccs[0]
        secs = cc.sections()
        self.assertEqual(secs[7].track.name, 'Dancin')

    def testCloudcastsDescription(self):
        self.mc.register_cloudcast(spartacus, partytime)
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        cc = ccs[0]
        self.assertEqual(cc.description(), 'Bla bla')

    def testLimit(self):
        self.mc.register_cloudcasts(spartacus, [partytime, lambiance])
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        self.assertEqual(len(ccs), 2)
        ccs = u.cloudcasts(limit=1)
        self.assertEqual(len(ccs), 1)
        self.assertEqual(ccs[0].key, 'party-time')
        ccs = u.cloudcasts(offset=1)
        self.assertEqual(len(ccs), 1)
        self.assertEqual(ccs[0].key, 'lambiance')

    def testYaml(self):
        self.mc.i_am(spartacus)
        self.mc.mock_upload(self.m.me())
        mp3file = io.StringIO(u'\x00' * 30)
        with open('example.yml') as f:
            self.m.upload_yml_file(f, mp3file)
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        self.assertEqual(len(ccs), 1)
        cc = ccs[0]
        sections = cc.sections()
        self.assertEqual(len(sections), 16)
        section = sections[6]
        self.assertEqual(section.start_time, 688)
        self.assertEqual(section.track.artist.name, 'Menelik & No Se')
        self.assertEqual(section.track.name, 'Quelle aventure')
        tags = ['Sample chain', 'Samples', 'Hip hop', 'Pop']
        self.assertEqual(cc.tags, tags)
        self.assertIn("In this mix we jump", cc.description())

    def testOauthUrl(self):
        full_url = self.o.authorize_url()
        # Check URL without parameters.
        self.assertEqual(
            full_url.split('?')[0], mixcloud.OAUTH_ROOT + '/authorize')
        # Check query string parameters.
        u = urlsplit(full_url)
        params = parse_qs(u.query)
        self.assertEqual(params['client_id'][0], self.client_id)
        self.assertEqual(params['redirect_uri'][0], self.redirect_uri)

    def testOauthExchange(self):
        self.mc.oauth_exchange()
        access_token = self.o.exchange_token('my_code')
        self.assertEqual(access_token, 'my_access_token')

    def testOauthExchangeServerFail(self):
        self.mc.oauth_exchange_fail()
        with self.assertRaises(mixcloud.MixcloudOauthError):
            self.o.exchange_token('my_code')

    @mock.patch('netrc.netrc.authenticators')
    @mock.patch('netrc.netrc.__init__')
    def testNetrc(self, netrc_init, netrc_authenticators):
        netrc_init.return_value = None  # Don't blow up when netrc missing.
        netrc_authenticators.return_value = ('', None, 'my_access_token')

        m = mixcloud.Mixcloud()
        self.assertEqual(m.access_token, 'my_access_token')


tests/tests.py
# FIXME: TESTS!!!

import csv
import datetime
import io
import unittest
from unittest import mock
from urllib.parse import parse_qs, urlsplit

import dateutil.tz
import httpretty

import mixcloud3 as mixcloud
from mixcloud3.mock import MockServer, parse_headers, parse_multipart


def parse_tracklist(s):
    s = s.split('\n')[1:-1]
    reader = csv.reader(s, delimiter='|', quoting=csv.QUOTE_NONE)
    t = [mixcloud.Section(int(line[0]),
                          mixcloud.Track(line[1].strip(),
                                         mixcloud.Artist(None,
                                                         line[2].strip()
                                                         )
                                         )
                          )
         for line in reader]
    return t


afx = mixcloud.Artist('aphex-twin', 'Aphex Twin')
spartacus = mixcloud.User('spartacus', 'Spartacus')
partytime = mixcloud.Cloudcast(
    'party-time', 'Party Time',
    parse_tracklist("""
       0 | Samurai (12" Mix)              | Jazztronik
     416 | Refresher                      | Time of your life
     716 | My time (feat. Crystal Waters) | Dutch
    1061 | Definition of House            | Minimal Funk
    1500 | I dont know                    | Mint Royale
    1763 | Thrill Her                     | Michael Jackson
    2123 | Happy (feat.Charlise)          | Elio Isola
    2442 | Dancin                         | Erick Morillo et al
    2738 | All in my head                 | Kosheen
    """),
    ['Funky house', 'Funk', 'Soul'],
    'Bla bla',
    spartacus,
    datetime.datetime(2009, 8, 2, 16, 55, 1, tzinfo=dateutil.tz.tzutc()),
)
lambiance = mixcloud.Cloudcast(
    'lambiance', "L'ambiance",
    parse_tracklist("""
     10 | As Serious As Your Life                     | Four Tet
     20 | Dynamic Symmetry                            | BT
     30 | Vessel                                      | Jon Hopkins
     40 | Vordhosbn                                   | Aphex Twin
     50 | Colour Eye                                  | Jon Hopkins
     60 | Flite                                       | The Cinematic Orchestra
     70 | Altibzz                                     | Autechre
     80 | Untitled [SAW2 CD1 Track1] (Four Tet remix) | Aphex Twin
     90 | Angelica                                    | Lamb
    100 | Quixotic                                    | Spartacus
    110 | Monday - Paracetamol                        | Ulrich Schnauss
    120 | Aquarius                                    | Boards of Canada
    130 | Channel 1 Suite                             | The Cinematic Orchestra
    """),
    ['Idm', 'Originals', 'Ambient'],
    'Bla bla bla',
    spartacus,
    datetime.datetime(2010, 3, 11, 21, 53, 8, tzinfo=dateutil.tz.tzutc()),
)


class TestMixcloud(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        httpretty.enable()

    def setUp(self):
        self.client_id = 'qwerty'
        self.client_secret = 'ytrewq'
        self.redirect_uri = 'http://localhost/mixcloud-callback'
        self.o = mixcloud.MixcloudOauth(
            client_id=self.client_id, client_secret=self.client_secret,
            redirect_uri=self.redirect_uri)
        self.m = mixcloud.Mixcloud()
        httpretty.reset()
        self.mc = MockServer()

    def testArtist(self):
        self.mc.register_artist(afx)
        resp = self.m.artist('aphex-twin')
        self.assertEqual(resp.name, 'Aphex Twin')

    def testUser(self):
        self.mc.register_user(spartacus)
        resp = self.m.user('spartacus')
        self.assertEqual(resp.name, 'Spartacus')

    def testCloudcast(self):
        self.mc.register_cloudcast(spartacus, partytime)
        resp = self.m.user('spartacus')
        cc = resp.cloudcast('party-time')
        self.assertEqual(cc.name, 'Party Time')
        secs = cc.sections()
        self.assertEqual(len(secs), 9)
        sec = secs[1]
        self.assertEqual(sec.start_time, 416)
        self.assertEqual(sec.track.name, 'Refresher')
        self.assertEqual(sec.track.artist.name, 'Time of your life')
        self.assertEqual(cc.tags, ['Funky house', 'Funk', 'Soul'])
        self.assertEqual(cc.description(), 'Bla bla')
        self.assertTrue(cc.picture().startswith('http'))
        self.assertEqual(cc.user.key, 'spartacus')
        self.assertEqual(cc.user.name, 'Spartacus')

    def testCloudcasts(self):
        self.mc.register_cloudcast(spartacus, partytime)
        resp = self.m.user('spartacus')
        ccs = resp.cloudcasts()
        self.assertEqual(len(ccs), 1)
        cc = ccs[0]
        self.assertEqual(cc.name, 'Party Time')

    def testLogin(self):
        self.mc.i_am(spartacus)
        user = self.m.me()
        self.assertEqual(user.key, spartacus.key)
        self.assertEqual(user.name, spartacus.name)

    def testUpload(self):
        self.mc.i_am(spartacus)

        def upload_callback(request, uri, headers):
            data = parse_multipart(request.body)
            self.assertIn('mp3', data)
            self.assertIn('picture', data)
            name = data['name']
            key = mixcloud.slugify(name)
            sections, tags = parse_headers(data)
            description = data['description']
            me = self.m.me()
            created_time = datetime.datetime.now()
            cc = mixcloud.Cloudcast(key, name, sections, tags,
                                    description, me, created_time)
            self.mc.register_cloudcast(me, cc)
            return (200, headers, '{}')

        self.mc.handle_upload(upload_callback)
        mp3file = io.StringIO(u'\x00' * 30)
        picturefile = io.StringIO(u'\x00' * 30)
        r = self.m.upload(partytime, mp3file, picturefile=picturefile)
        self.assertEqual(r.status_code, 200)
        me = self.m.me()
        cc = me.cloudcast('party-time')
        self.assertEqual(cc.name, 'Party Time')
        secs = cc.sections()
        self.assertEqual(len(secs), 9)
        sec = secs[3]
        self.assertEqual(sec.start_time, 1061)
        self.assertEqual(sec.track.name, 'Definition of House')
        self.assertEqual(sec.track.artist.name, 'Minimal Funk')
        self.assertEqual(cc.tags, ['Funky house', 'Funk', 'Soul'])
        self.assertEqual(cc.description(), 'Bla bla')

    def testCloudcastsSection(self):
        self.mc.register_cloudcast(spartacus, partytime)
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        cc = ccs[0]
        secs = cc.sections()
        self.assertEqual(secs[7].track.name, 'Dancin')

    def testCloudcastsDescription(self):
        self.mc.register_cloudcast(spartacus, partytime)
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        cc = ccs[0]
        self.assertEqual(cc.description(), 'Bla bla')

    def testLimit(self):
        self.mc.register_cloudcasts(spartacus, [partytime, lambiance])
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        self.assertEqual(len(ccs), 2)
        ccs = u.cloudcasts(limit=1)
        self.assertEqual(len(ccs), 1)
        self.assertEqual(ccs[0].key, 'party-time')
        ccs = u.cloudcasts(offset=1)
        self.assertEqual(len(ccs), 1)
        self.assertEqual(ccs[0].key, 'lambiance')

    def testYaml(self):
        self.mc.i_am(spartacus)
        self.mc.mock_upload(self.m.me())
        mp3file = io.StringIO(u'\x00' * 30)
        with open('tests/example.yml') as f:
            self.m.upload_yml_file(f, mp3file)
        u = self.m.user('spartacus')
        ccs = u.cloudcasts()
        self.assertEqual(len(ccs), 1)
        cc = ccs[0]
        sections = cc.sections()
        self.assertEqual(len(sections), 16)
        section = sections[6]
        self.assertEqual(section.start_time, 688)
        self.assertEqual(section.track.artist.name, 'Menelik & No Se')
        self.assertEqual(section.track.name, 'Quelle aventure')
        tags = ['Sample chain', 'Samples', 'Hip hop', 'Pop']
        self.assertEqual(cc.tags, tags)
        self.assertIn("In this mix we jump", cc.description())

    def testOauthUrl(self):
        full_url = self.o.authorize_url()
        # Check URL without parameters.
        self.assertEqual(
            full_url.split('?')[0], mixcloud.OAUTH_ROOT + '/authorize')
        # Check query string parameters.
        u = urlsplit(full_url)
        params = parse_qs(u.query)
        self.assertEqual(params['client_id'][0], self.client_id)
        self.assertEqual(params['redirect_uri'][0], self.redirect_uri)

    def testOauthExchange(self):
        self.mc.oauth_exchange()
        access_token = self.o.exchange_token('my_code')
        self.assertEqual(access_token, 'my_access_token')

    def testOauthExchangeServerFail(self):
        self.mc.oauth_exchange_fail()
        with self.assertRaises(mixcloud.MixcloudOauthError):
            self.o.exchange_token('my_code')

    @mock.patch('netrc.netrc.authenticators')
    @mock.patch('netrc.netrc.__init__')
    def testNetrc(self, netrc_init, netrc_authenticators):
        netrc_init.return_value = None  # Don't blow up when netrc missing.
        netrc_authenticators.return_value = ('', None, 'my_access_token')

        m = mixcloud.Mixcloud()
        self.assertEqual(m.access_token, 'my_access_token')


mixcloud3/__init__.py
import collections
import datetime
import netrc
from dataclasses import dataclass
from typing import Dict, List, Optional
from urllib.parse import urlencode

import dateutil.parser
import requests
import yaml
from slugify import slugify

from utils import logger

NETRC_MACHINE = 'mixcloud-api'
API_ROOT = 'https://api.mixcloud.com'
OAUTH_ROOT = 'https://www.mixcloud.com/oauth'

API_ERROR_MESSAGE = "Mixcloud {} API returned HTTP code {}"

log = logger(__name__)


class MixcloudOauthError(Exception):
    pass


class APIError(Exception):
    pass


def get(*args, **kwargs):
    """A wrapper for requests.GET method"""
    response = requests.get(*args, **kwargs)
    if response.status_code == 200:
        return response
    raise APIError(API_ERROR_MESSAGE.format('GET', response.status_code))


def post(*args, **kwargs):
    """A wrapper for requests.POST method"""
    response = requests.post(*args, **kwargs)
    if response.status_code == 200:
        return response
    raise APIError(API_ERROR_MESSAGE.format('POST', response.status_code))


def setup_yaml():
    def construct_yaml_str(self, node):
        # Override the default string handling function
        # to always return unicode objects
        return self.construct_scalar(node)
    tag = u'tag:yaml.org,2002:str'
    yaml.Loader.add_constructor(tag, construct_yaml_str)
    yaml.SafeLoader.add_constructor(tag, construct_yaml_str)


def get_many(url, limit=None, offset=None):
    """Gets many records from Mixcloud API"""
    params = {}
    if limit is not None:
        params['limit'] = limit
    if offset is not None:
        params['offset'] = offset
    r = get(url, params=params)
    return r.json()


def get_all(url):
    """A wrapper for `get_many()`: a generator getting and iterating through all results"""
    data = get_many(url, limit=50)
    yield from data['data']
    while 'next' in data['paging']:
        data = get_many(data['paging']['next'])
        yield from data['data']


class MixcloudOauth:
    """
    Assists in the OAuth dance with Mixcloud to get an access token.
    """

    def __init__(self, client_id=None, client_secret=None, redirect_uri=None):
        self.client_id = client_id
        self.client_secret = client_secret
        self.redirect_uri = redirect_uri

    def authorize_url(self):
        """
        Return a URL to redirect the user to for OAuth authentication.
        """
        auth_url = OAUTH_ROOT + '/authorize'
        params = {
            'client_id': self.client_id,
            'redirect_uri': self.redirect_uri,
        }
        return "{}?{}".format(auth_url, urlencode(params))

    def exchange_token(self, code):
        """
        Exchange the authorization code for an access token.
        """
        access_token_url = OAUTH_ROOT + '/access_token'
        params = {
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'redirect_uri': self.redirect_uri,
            'code': code,
        }
        resp = requests.get(access_token_url, params=params)
        if not resp.ok:
            raise MixcloudOauthError("Could not get access token.")
        return resp.json()['access_token']


class Mixcloud:

    def __init__(self, api_root=API_ROOT, access_token=None):
        self.api_root = api_root
        if access_token is None:
            try:
                # Check there is a netrc file.
                netrc_auth = netrc.netrc()
            except FileNotFoundError:
                pass
            else:
                try:
                    # Attempt netrc lookup.
                    credentials = netrc_auth.authenticators(NETRC_MACHINE)
                    if netrc_auth:
                        access_token = credentials[2]
                except netrc.NetrcParseError:
                    # Configuration errors unrelated to the Mixcloud entry
                    # will cause this exception to be thrown, whether or not
                    # there is a Mixcloud entry.
                    pass
        self.access_token = access_token

    def artist(self, key):
        url = '{}/artist/{}'.format(self.api_root, key)
        r = get(url)
        return Artist.from_json(r.json())

    def user(self, key):
        url = '{}/{}'.format(self.api_root, key)
        r = get(url)
        return User.from_json(r.json(), m=self)

    def me(self):
        url = '{}/me/'.format(self.api_root)
        r = get(url, {'access_token': self.access_token})
        return User.from_json(r.json(), m=self)

    def upload(self, cloudcast, mp3file, picturefile=None):
        url = '{}/upload/'.format(self.api_root)
        payload = {'name': cloudcast.name,
                   'percentage_music': 100,
                   'description': cloudcast.description(),
                   }
        for num, sec in enumerate(cloudcast.sections()):
            payload['sections-%d-artist' % num] = sec.track.artist.name
            payload['sections-%d-song' % num] = sec.track.name
            payload['sections-%d-start_time' % num] = sec.start_time

        for num, tag in enumerate(cloudcast.tags):
            payload['tags-%s-tag' % num] = tag

        files = {'mp3': mp3file}
        if picturefile is not None:
            files['picture'] = picturefile

        r = post(url, data=payload, params={'access_token': self.access_token}, files=files)
        return r

    def upload_yml_file(self, ymlfile, mp3file):
        user = self.me()
        cloudcast = Cloudcast.from_yml(ymlfile, user)
        _ = self.upload(cloudcast, mp3file)


@dataclass
class Artist:

    key: str
    name: str

    m: Optional[Mixcloud] = None

    @staticmethod
    def from_json(data):
        return Artist(data['slug'], data['name'])

    @staticmethod
    def from_yml(artist):
        return Artist(slugify(artist), artist)


@dataclass
class User:

    key: str
    name: str
    m: Optional[Mixcloud] = None

    _metadata: Optional[Dict] = None

    @staticmethod
    def from_json(data, m=None):
        if 'username' in data and 'name' in data:
            return User(data['username'], data['name'], m=m)

    def __repr__(self):
        return "<User:{}>".format(self.name)

    def __str__(self):
        return repr(self)

    def _get_metadata(self):
        url = '{}/{}/?metadata=1'.format(self.m.api_root, self.name)
        r = get(url)
        data = r.json()
        return data['metadata']['connections']

    def cloudcast(self, key):
        url = '{}/{}/{}'.format(self.m.api_root, self.key, key)
        r = get(url)
        data = r.json()
        return Cloudcast.from_json(data)

    def cloudcasts(self, limit=None, offset=None, all=False):
        data = get_many('{}/{}/cloudcasts/'.format(self.m.api_root, self.key), limit, offset)
        return [Cloudcast.from_json(d, m=self.m) for d in data['data']]

    def playlist(self, key):
        r = get('{}/{}/playlists/{}'.format(self.m.api_root, self.key, key))
        data = r.json()
        return Playlist.from_json(data, m=self.m)

    def playlists(self):
        pl = self.metadata.get("playlists")
        if pl:
            for playlist in get_all(pl):
                yield Playlist.from_json(playlist)

    @property
    def metadata(self):
        if not self._metadata:
            self._metadata = self._get_metadata()
        return self._metadata


@dataclass
class Playlist:

    key: str
    url: str
    name: str
    owner: str
    slug: str
    cloudcast_count: Optional[int] = 0
    created_time: Optional[datetime.datetime] = None
    updated_time: Optional[datetime.datetime] = None

    m: Optional[Mixcloud] = None

    def cloudcasts(self, limit=None, offset=None, all=False):
        url = '{}{}cloudcasts'.format(API_ROOT, self.key)
        if all:
            data = get_all(url)
        else:
            data = get_many(url, limit=limit, offset=offset)
        for cast in data:
            yield Cloudcast.from_json(cast, m=self.m)

    @staticmethod
    def from_json(d, m=None):
        ctime = dateutil.parser.parse(d['created_time']) if 'created_time' in d else None
        mtime = dateutil.parser.parse(d['updated_time']) if 'updated_time' in d else None
        return Playlist(
            d['key'],
            d['url'],
            d['name'],
            User.from_json(d['owner']),
            d['slug'],
            d.get('cloudcast_count', 0),
            ctime,
            mtime,
            m=m
        )


@dataclass
class Cloudcast:

    key: str
    url: str
    name: str
    tags: Optional[List['Tag']] = None
    created_time: Optional[datetime.datetime] = None
    updated_time: Optional[datetime.datetime] = None
    play_count: Optional[int] = 0
    favorite_count: Optional[int] = 0
    comment_count: Optional[int] = 0
    listener_count: Optional[int] = 0
    repost_count: Optional[int] = 0
    pictures: Optional[Dict] = None
    slug: Optional[int] = None
    user: Optional[User] = None
    hidden_stats: Optional[bool] = None
    audio_length: Optional[int] = 0

    _description: Optional[str] = ''
    _sections: Optional[List['Section']] = None

    m: Optional[Mixcloud] = None

    @staticmethod
    def from_json(d, m=None):
        if 'sections' in d:
            sections = Section.list_from_json(d['sections'])
        else:
            sections = None
        desc = d.get('description')
        tags = Tag.list_from_json(d['tags'])
        user = User.from_json(d['user'])
        created_time = dateutil.parser.parse(d['created_time'])
        updated_time = dateutil.parser.parse(d['updated_time'])
        pictures = d.get('pictures')
        return Cloudcast(
            d['key'],
            d['url'],
            d['name'],
            tags,
            created_time,
            updated_time,
            d.get('play_count'),
            d.get('favorite_count'),
            d.get('comment_count'),
            d.get('listener_count'),
            d.get('repost_count'),
            pictures,
            d['slug'],
            user,
            d.get('hidden_stats'),
            d['audio_length'],
            desc,
            sections,
            m
        )

    def _load(self):
        url = '{}{}'.format(self.m.api_root, self.key)
        r = get(url)
        d = r.json()
        self._sections = Section.list_from_json(d['sections'])
        self._description = d['description']

    @property
    def sections(self):
        """
        Depending on the data available when the instance was created,
        it may be necessary to fetch data.
        """
        if self._sections is None:
            self._load()
        return self._sections

    @property
    def description(self):
        """
        May hit server. See Cloudcast.sections
        """
        if self._description is None:
            self._load()
        return self._description

    @property
    def picture(self):
        return self.pictures['large']

    @staticmethod
    def from_yml(f, user):
        setup_yaml()
        d = yaml.load(f, Loader=yaml.FullLoader)
        name = d['title']
        sections = [Section.from_yml(s) for s in d['tracks']]
        key = slugify(name)
        tags = d['tags']
        description = d['desc']
        created_time = None
        c = Cloudcast(key, name, sections, tags, description,
                      user, created_time)
        return c


# TODO: refactor to dataclass
class Section(collections.namedtuple('_Section', 'start_time track')):

    @staticmethod
    def from_json(d):
        return Section(d['start_time'], Track.from_json(d['track']))

    @staticmethod
    def list_from_json(d):
        return [Section.from_json(s) for s in d]

    @staticmethod
    def from_yml(d):
        artist = Artist.from_yml(d['artist'])
        track = d['track']
        return Section(d['start'], Track(track, artist))


@dataclass
class Track:

    name: str
    artist: Artist

    @staticmethod
    def from_json(d):
        return Track(d['name'], Artist.from_json(d['artist']))


@dataclass
class Tag:
    key: str
    url: str
    name: str

    @staticmethod
    def from_json(d):
        return Tag(d['key'], d['url'], d['name'])

    @staticmethod
    def list_from_json(d):
        return [Tag.from_json(t) for t in d]

    def __str__(self):
        return self.name

    def __repr__(self):
        return str(self)
